File: datasets/Python/Copilot/DataStructures/medium4.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.count = 0

class Contacts:
    def __init__(self):
        self.root = TrieNode()

    def add(self, name):
        node = self.root
        for char in name:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
            node.count += 1

    def find(self, partial):
        node = self.root
        for char in partial:
            if char not in node.children:
                return 0
            node = node.children[char]
        return node.count

# Manual Input
n = int(input("Enter number of operations: "))
contacts = Contacts()
results = []
for _ in range(n):
    parts = input().split()
    if parts[0] == 'add':
        contacts.add(parts[1])
    elif parts[0] == 'find':
        results.append(contacts.find(parts[1]))

print("Results:")
for res in results:
    print(res)

File: datasets/C++/Copilot/DataStructures/medium4.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

struct TrieNode {
    int count = 0;
    unordered_map<char, TrieNode*> children;
};

class Contacts {
    TrieNode* root;
public:
    Contacts() { root = new TrieNode(); }

    void add(const string& name) {
        TrieNode* node = root;
        for (char c : name) {
            if (!node->children.count(c))
                node->children[c] = new TrieNode();
            node = node->children[c];
            node->count++;
        }
    }

    int find(const string& partial) {
        TrieNode* node = root;
        for (char c : partial) {
            if (!node->children.count(c)) return 0;
            node = node->children[c];
        }
        return node->count;
    }
};

int main() {
    int n;
    cout << "Enter number of operations: ";
    cin >> n;
    cin.ignore();
    Contacts contacts;
    vector<int> results;

    for (int i = 0; i < n; ++i) {
        string op, arg;
        getline(cin, op);
        size_t space = op.find(' ');
        string cmd = op.substr(0, space);
        string val = op.substr(space + 1);
        if (cmd == "add") contacts.add(val);
        else if (cmd == "find") results.push_back(contacts.find(val));
    }

    cout << "Results:\n";
    for (int r : results) cout << r << "\n";
    return 0;
}
